### PR TITLE
Bounds

### PR DESCRIPTION
**Pull request by BibMartin**
Not finished. Adresses #139

### CODE REVIEW

File: folium/utilities.py [hunk 1]
```diff
@@ -428,3 +428,37 @@ def _locations_tolist(x):
...
+                out = []
+                for y in x:
+                    out += iter_points(y)
+                return out
+            else:
+                return [x]
+        else:
+            return []
+    else:
+        raise ValueError('List type expected {!r}'.format(x))
```
Comment (ocefpaf) [inline folium/utilities.py]: Maybe it is better: `List type expected.  Got {!r}.`

File: folium/utilities.py [hunk 2]
```diff
@@ -428,3 +428,37 @@ def _locations_tolist(x):
...
+    else:
+        return max(x,y)
+
+def iter_points(x):
+    """Iterates over a list representing a feature, and returns a list of points,
+    whatever the shape of the array (Point, MultiPolyline, etc).
+    """
+    if isinstance(x,list):
+        if len(x):
+            if isinstance(x[0],list):
```
Comment (ocefpaf) [inline folium/utilities.py]: space after the `,` :ghost:

File: folium/map.py [hunk 3]
```diff
@@ -421,6 +417,11 @@ def __init__(self, location, popup=None, icon=None):
             {% endmacro %}
             """)
 
+    def _get_self_bounds(self):
+        """Computes the bounds of the object itself (not including it's children)
+        in the form [[lat_min, lon_min], [lat_max, lon_max]]
+        """
+        return [[self.location[0],self.location[1]],[self.location[0],self.location[1]]]
```
Comment (ocefpaf) [inline folium/map.py]: Spaces after the `,`

File: folium/features.py [hunk 4]
```diff
@@ -616,6 +669,23 @@ def __init__(self, locations, color=None, weight=None,
                 {{this._parent.get_name()}}.addLayer({{this.get_name()}});
             {% endmacro %}
             """)  # noqa
+    def _get_self_bounds(self):
+        """Computes the bounds of the object itself (not including it's children)
+        in the form [[lat_min, lon_min], [lat_max, lon_max]]
+        """
+        bounds = [[None,None],[None,None]]
```
Comment (ocefpaf) [inline folium/features.py]: Spaces...

File: folium/features.py [hunk 5]
```diff
@@ -616,6 +669,23 @@ def __init__(self, locations, color=None, weight=None,
...
+        for point in iter_points(self.data):
+            bounds = [
+                [
+                    none_min(bounds[0][0], point[0]),
```
Comment (ocefpaf) [inline folium/features.py]: I see what you are trying to do here (and I do not have a better suggestion right now), but it looks like we could do something simpler...

**Comment by ocefpaf** [comment]
Thanks @BibMartin! I made a few minor comments.  Let me know once you are done so I can review it again.

File: folium/utilities.py (hunk 1 above)
Comment (BibMartin) [inline folium/utilities.py]: ok, I'll do it

**Comment by BibMartin** [comment]
Thanks @ocefpaf for first review. I'm okay with your comments and will apply them asap.
Before merging, I would like to:

* add `get_bounds` call in most tests (see `test_topo_json` for example)
* plug back something *like* `Map._auto_bounds` that uses this method.

**Comment by BibMartin** [comment]
@ocefpaf 
I'm finished with this one. After rethinking, I guess it's no need to re-create the `_auto_bounds` method, provided everyone can create his own bounds in doing :

    map.fit_bounds(my_feature.get_bounds())

We can eventually add something in `Map.render` so that if no `FitBounds` has been defined, we do

    map.fit_bounds(map.get_bounds())

Tell me if you want I add it before squashing.

**Comment by BibMartin** [comment]
@themiurgo I would be interested to have your point of view also ; I realize this addresses #244 in the same time as #139 .

**Comment by ocefpaf** [comment]
Awesome!  

http://nbviewer.ipython.org/github/ocefpaf/folium_notebooks/blob/master/test_fit_bounds.ipynb

Thanks @BibMartin!!!

**Comment by themiurgo** [comment]
Sorry, coming late for this. Looks great though! :)

**Pull request merged**